src/imitation/data/types.py

```diff
@@ -22,6 +27,177 @@
 T = TypeVar("T")
 
 AnyPath = Union[str, bytes, os.PathLike]
+
+
+@dataclasses.dataclass(frozen=True)
+class DictObs:
+    """Stores observations from an environment with a dictionary observation space.
+
+    Provides an interface that is similar to observations in a numpy array.
+    Length, slicing, indexing, and iterating operations will operate on the first
+    dimension of the constituent arrays, as they would for observations in a single
+    array.
+
+    There are also utility functions for mapping / stacking / concatenating
+    lists of dictobs.
+    """
+
+    d: Dict[str, np.ndarray]
+
+    @classmethod
+    def from_obs_list(cls, obs_list: List[Dict[str, np.ndarray]]):
```

Comment:
Nit: check that all elements have the same keys? (This will error if `obs_list[i]` contains a key not present in `obs_list[0]` but not vice-versa.)